examples/Custom_Hooks/basics.py

```diff
@@ -0,0 +1,108 @@
+## Registering Custom Hooks
+
+# This example demonstrates the basic mechanics of the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility,
+# which lets you hook into any callable of your choice:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+
+from dataclasses import dataclass
+from time import perf_counter
+
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import RandomRecommender
+from baybe.searchspace import SearchSpace
+from baybe.utils.basic import register_hooks
+
+### Defining the Hooks
+
+# We start by defining a simple hook that lets us inspect the names of the parameters
+# involved in the recommendation process.
+# For this purpose, we match its signature to that of
+# {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`:
+#
+
+# ```{admonition} Signature components
+# :class: important
+# Note that, if provided, annotations must match **exactly** those of the target signature.
+# However, annotations are completely optional
+# — only the names/order of the signature parameters and their defaults matter.
+#
+# For the sake of demonstration, we only provide the annotation for the
+# relevant `searchspace` parameter of the hook in the example below.
+# ```
+
+
+def print_parameter_names_hook(
+    self,
+    batch_size,
+    searchspace: SearchSpace,
+    objective=None,
+    measurements=None,
+):
+    """Print the names of the parameters spanning the search space."""
+    print(f"Search space parameters: {[p.name for p in searchspace.parameters]}")
+
+
+# Additionally, we set up a class that provides a combination of hooks for measuring
+# the time needed to compute the recommendations:
+
+
+@dataclass
+class ElapsedTimePrinter:
+    """Helper class for measuring the time between two calls."""
+
+    last_call_time: float | None = None
+
+    def start(
+        instance,
+        self,
+        batch_size,
+        searchspace,
+        objective=None,
+        measurements=None,
+    ):
+        """Start the timer."""
+        instance.last_call_time = perf_counter()
+
+    def measure(
+        instance,
+        self,
```

Comment:
its indeed a bit confusing to have `intance` and `self` here, why is that?

Reply:
It's actually quite simple: the **method** (i.e. not function) we are patching of course has "self" as first argument. The plain logic that is implemented so far goes via names, which is required for matching if you want to be able to omit parameters. But: the hook itself is also a method, which needs to have the special "bound" parameter (i.e. the first arg). So in conclusion:
* The second parameter must be called `self`, for the match with the target
* Hence, we need to use a different name for the first one (remember that "self" is just a convention, so we can replace it with anything we want).

In the updated example, this no longer occurs because "self" is now just omitted (since not used), but the problem as such still remains. We can work around it though, by implementing a special case, name by considering if or not the target is a method and then apply exceptions. For instance:
* We can allow that "self" from target can be named arbitrarily in the hook but must always comes the first/second argument, depending on whether or not the hook itself is a method.
* We could further allow that, for this special case, the hook annotation can be arbitrary and does not need to match the annotation from the target (since "self" is never annotated in methods) <-- this is what I tried to explain in the other thread.

However, this would require quite a bit of additional logic. What do you think, is it necessary? Please have a look at the current version and let me know. So far, I haven't updated the tests and will wait until we have agreed on what to do here.

Reply:
After having spent the entire afternoon of this, I really don't want invest any more time into it. The whole mechanism turns out way more complicated than expected when you [start thinking](https://github.com/emdgroup/baybe/pull/247#discussion_r1639926564) about the difference between methods/functions. So I'd say let's stick with the current functionality for now. What is your opinion?

Reply:
agreed that the latest version is good enough